zetaclient/bitcoin/fee.go

```diff
@@ -0,0 +1,234 @@
+	bytes1stWitness         = 110 // the 1st witness incurs about 110 bytes and it may vary
+	bytesPerWitness         = 108 // each additional witness incurs about 108 bytes and it may vary
+	defaultDepositorFeeRate = 20  // 20 sat/byte is the default depositor fee rate
+
+	outTxBytesMin = uint64(239)  // 239vB == EstimateSegWitTxSize(2, 2, toP2WPKH)
+	outTxBytesMax = uint64(1543) // 1543v == EstimateSegWitTxSize(21, 2, toP2TR)
+	outTxBytesAvg = uint64(245)  // 245vB is a suggested gas limit for zeta core
+)
+
+var (
+	BtcOutTxBytesDepositor  uint64
+	BtcOutTxBytesWithdrawer uint64
+	DefaultDepositorFee     float64
+)
+
+func init() {
+	BtcOutTxBytesDepositor = OuttxSizeDepositor()   // 68vB, the outtx size incurred by the depositor
+	BtcOutTxBytesWithdrawer = OuttxSizeWithdrawer() // 177vB, the outtx size incurred by the withdrawer
+
+	// default depositor fee calculation is based on a fixed fee rate of 20 sat/byte just for simplicity.
+	// In reality, the fee rate on UTXO deposit is different from the fee rate when the UTXO is spent.
+	DefaultDepositorFee = DepositorFee(defaultDepositorFeeRate) // 0.00001360 (20 * 68vB / 100000000)
+}
+
+// FeeRateToSatPerByte converts a fee rate in BTC/KB to sat/byte.
+func FeeRateToSatPerByte(rate float64) *big.Int {
+	// #nosec G701 always in range
+	satPerKB := new(big.Int).SetInt64(int64(rate * btcutil.SatoshiPerBitcoin))
+	return new(big.Int).Div(satPerKB, big.NewInt(bytesPerKB))
+}
+
+// WiredTxSize calculates the wired tx size in bytes
+func WiredTxSize(numInputs uint64, numOutputs uint64) uint64 {
+	// Version 4 bytes + LockTime 4 bytes + Serialized varint size for the
+	// number of transaction inputs and outputs.
+	// #nosec G701 always positive
+	return uint64(8 + wire.VarIntSerializeSize(numInputs) + wire.VarIntSerializeSize(numOutputs))
+}
+
+// EstimateOuttxSize estimates the size of a outtx in vBytes
+func EstimateOuttxSize(numInputs uint64, payees []btcutil.Address) uint64 {
+	if numInputs == 0 {
+		return 0
+	}
+	// #nosec G701 always positive
+	numOutputs := 2 + uint64(len(payees))
+	bytesWiredTx := WiredTxSize(numInputs, numOutputs)
+	bytesInput := numInputs * bytesPerInput
+	bytesOutput := uint64(2) * bytesPerOutputP2WPKH // new nonce mark, change
+
+	// calculate the size of the outputs to payees
+	bytesToPayees := uint64(0)
+	for _, to := range payees {
+		bytesToPayees += GetOutputSizeByAddress(to)
+	}
+	// calculate the size of the witness
+	bytesWitness := bytes1stWitness + (numInputs-1)*bytesPerWitness
+	// https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#transaction-size-calculations
+	// Calculation for signed SegWit tx: blockchain.GetTransactionWeight(tx) / 4
+	return bytesWiredTx + bytesInput + bytesOutput + bytesToPayees + bytesWitness/blockchain.WitnessScaleFactor
+}
+
+// GetOutputSizeByAddress returns the size of a tx output in bytes by the given address
+func GetOutputSizeByAddress(to btcutil.Address) uint64 {
```

Comment:
This conversion check is very nice, might be able to use this to determine address type in common folder.

Reply:
Do you mean the common folder in `zetaclient`? It now only contains logger and constant (gas multiplier) that used across chains. This Bitcoin tx script and size stuff is only needed for Bitcoin so maybe we just leave it here?

Reply:
I meant zeta-node/common/bitcoin/

I just thought it might be useful in the address_taproot file.

Reply:
IC. The `address_taproot.go` was actually a copied file from latest version of `btcsuite` package because we're using the old version which doesn't have taproot support.